MatrixSDK/VoIP/CallKit/MXCallKitAdapter.m

```diff
@@ -0,0 +1,285 @@
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import "MXCallKitAdapter.h"
+
+@import AVFoundation;
+@import CallKit;
+@import UIKit;
+
+#import "MXCall.h"
+#import "MXJingleCallAudioSessionConfigurator.h"
+#import "MXUser.h"
+#import "MXSession.h"
+
+@interface MXCallKitAdapter () <CXProviderDelegate>
+
+@property (nonatomic) CXProvider *provider;
+@property (nonatomic) CXCallController *callController;
+
+@property (nonatomic) NSMutableDictionary<NSUUID *, MXCall *> *calls;
+
+@end
+
+@implementation MXCallKitAdapter
+
+- (instancetype)init
+{
+    if (self = [super init])
+    {
+        _provider = [[CXProvider alloc] initWithConfiguration:[MXCallKitAdapter configuration]];
+        [_provider setDelegate:self queue:nil];
+        
+        _callController = [[CXCallController alloc] initWithQueue:dispatch_get_main_queue()];
+        
+        _calls = [NSMutableDictionary dictionary];
+    }
+    
+    return self;
+}
+
+- (void)dealloc
+{
+    // CXProvider instance must be invalidated otherwise it will be leaked
+    [_provider invalidate];
+}
+
++ (CXProviderConfiguration *)configuration
+{
+    NSString *appDisplayName = [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleDisplayName"];
+    
+    CXProviderConfiguration *configuration = [[CXProviderConfiguration alloc] initWithLocalizedName:appDisplayName];
+    configuration.maximumCallGroups = 1;
+    configuration.maximumCallsPerCallGroup = 1;
+    configuration.supportsVideo = YES;
+    configuration.supportedHandleTypes = [NSSet setWithObject:@(CXHandleTypeGeneric)];
+//    configuration.iconTemplateImageData = UIImagePNGRepresentation([UIImage imageNamed:@"RiotCallKitLogo"]);
+    
+    return configuration;
+}
+
+#pragma mark - Public
+
+- (void)startCall:(MXCall *)call
+{
+    MXSession *mxSession = call.room.mxSession;
+    MXUser *caller = [mxSession userWithUserId:call.callerId];
+    NSUUID *callUUID = call.callUUID;
+    
+    NSString *handleValue;
+    NSString *contactIdentifier;
+    if (call.isConferenceCall)
+    {
+        handleValue = call.room.roomId;
+        contactIdentifier = call.room.state.displayname;
+    }
+    else
+    {
+        MXUser *callee = [mxSession userWithUserId:call.calleeId];
+        
+        handleValue = callee.userId;
+        contactIdentifier = callee.displayname;
+    }
+    
+    CXHandle *handle = [[CXHandle alloc] initWithType:CXHandleTypeGeneric value:handleValue];
+    CXStartCallAction *action = [[CXStartCallAction alloc] initWithCallUUID:callUUID handle:handle];
+    action.contactIdentifier = contactIdentifier;
+    
+    CXTransaction *transaction = [[CXTransaction alloc] initWithAction:action];
+    [self.callController requestTransaction:transaction completion:^(NSError * _Nullable error) {
+        if (error)
+            return;
```

Comment:
> at this time, is it still possible to back up and make the call within the app?
The call is already running by the time we are in this codepath.

I had a look at `CXErrorCodeRequestTransactionError`, this enum represent all possible errors and i don't thin we can encounter any of them, so we can ingnore an error here